mp/src/materialsystem/stdshaders/pbr_ps20b.fxc

```diff
@@ -0,0 +1,219 @@
+sampler EnvmapSampler			    : register(s2);		// Cubemap
+sampler ShadowDepthSampler		    : register(s4);		// Flashlight shadow depth map sampler
+sampler RandRotSampler			    : register(s5);		// RandomRotation sampler
+sampler FlashlightSampler		    : register(s6);		// Flashlight cookie 
+sampler LightmapSampler			    : register(s7);		// Lightmap
+sampler MRAOTextureSampler		    : register(s10);	// MRAO texture
+#if EMISSIVE
+sampler EmissionTextureSampler	    : register(s11);	// Emission texture
+#endif
+#if SPECULAR
+sampler SpecularTextureSampler		: register(s12);	// Specular F0 texture
+#endif
+
+#define ENVMAPLOD (g_EyePos.a)
+
+struct PS_INPUT
+{
+	float2 baseTexCoord				: TEXCOORD0;
+	float4 lightAtten				: TEXCOORD1;
+	float3 worldNormal				: TEXCOORD2;
+	float3 worldTangent				: TEXCOORD3;
+	float3 worldPos					: TEXCOORD4;
+	float3 projPos					: TEXCOORD5;
+	float4 lightmapTexCoord1And2	: TEXCOORD6;
+	float4 lightmapTexCoord3		: TEXCOORD7;
+};
+
+// Entry point
+float4 main(PS_INPUT i) : COLOR
+{
+
+#if PARALLAXOCCLUSION
+    //float3 viewDir = i.vViewTS;
+	//float2 correctedTexCoord = parallaxCorrect(i.baseTexCoord, viewDir , NormalTextureSampler , PARALLAX_HEIGHT , g_ParallaxParms.r , g_ParallaxParms.g);
+	//Debug Stuff - uncomment if testing shadow
+   	//correctedTexCoord = i.baseTexCoord; 
+	//float displayHeight = tex2D(MRAOTextureSampler, correctedTexCoord).a;
+#else
+	float2 correctedTexCoord = i.baseTexCoord;
+#endif
+
+#if USEENVAMBIENT
+	//float3 EnvAmbientCube[6];
+	//setupEnvMapAmbientCube(EnvAmbientCube);
+    #define EnvAmbientCube cAmbientCube
+#else
+	#define EnvAmbientCube cAmbientCube
+#endif
+
+	float4 albedo = tex2D(BaseTextureSampler, correctedTexCoord);
+	albedo.xyz *= g_BaseColor;
+
+	float3 mrao = tex2D(MRAOTextureSampler, correctedTexCoord).xyz;
+    float metalness = mrao.x, roughness = mrao.y, ambientOcclusion = mrao.z;
+#if EMISSIVE
+	float3 emission = tex2D(EmissionTextureSampler, correctedTexCoord).xyz;
+#endif
+#if SPECULAR
+	float3 specular = tex2D(SpecularTextureSampler, correctedTexCoord).xyz;
+#endif
+
+	//float3 surfNormal = normalize(i.worldNormal);
+	//float3 surfTangent;
+	//float3 surfBase;
+	//float flipSign;
+	//float3x3 normalBasis = compute_tangent_frame(surfNormal, i.worldPos, i.baseTexCoord, surfTangent, surfBase, flipSign);
+    float3x3 normalBasis = float3x3(i.worldTangent, cross(i.worldNormal, i.worldTangent), i.worldNormal);
+	float3 textureNormal = normalize((tex2D( NormalTextureSampler, correctedTexCoord).xyz - float3(0.5, 0.5, 0.5)) * 2);
+	float3 normal = normalize(mul(textureNormal, normalBasis)); // World Normal
+	//textureNormal.y *= flipSign; // Fixup textureNormal for ambient lighting
+
+	float3 outgoingLightDirection = normalize(g_EyePos.xyz - i.worldPos); // Lo
+	float lightDirectionAngle = max(0, dot(normal, outgoingLightDirection)); // cosLo
+
+	float3 specularReflectionVector = 2.0 * lightDirectionAngle * normal - outgoingLightDirection; // Lr
```

Comment:
Shouldn't most of this be in the common .h file as well?

Reply:
we can, right now i did the functions only since those will break the shaders the least.